apps/avatax/src/modules/avatax/calculate-taxes/avatax-product-line-calculate-taxes-factory.ts

```diff
@@ -0,0 +1,29 @@
+import { LineItemModel } from "avatax/lib/models/LineItemModel";
+import { TaxBaseLineFragment } from "generated/graphql";
+
+export class AvataxProductLineCalculateTaxesFactory {
+  createFromSaleorLine(args: {
+    saleorLine: TaxBaseLineFragment;
+    taxIncluded: boolean;
+    taxCode: string;
+    discounted: boolean;
+  }): LineItemModel {
+    return {
+      amount: args.saleorLine.totalPrice.amount,
+      taxIncluded: args.taxIncluded,
+      taxCode: args.taxCode,
+      quantity: args.saleorLine.quantity,
+      itemCode: this.getItemCodeFromSourceLine(args.saleorLine.sourceLine),
+      discounted: args.discounted,
+    };
+  }
+
+  private getItemCodeFromSourceLine(line: TaxBaseLineFragment["sourceLine"]) {
+    if (line.__typename === "CheckoutLine") {
+      return line.checkoutProductVariant.sku ?? line.checkoutProductVariant.id ?? "";
+    }
+    if (line.__typename === "OrderLine") {
+      return line.orderProductVariant?.sku ?? line.orderProductVariant?.id ?? "";
+    }
```

Comment:
suggestion: let's add invariant if there's some other __typename:

```suggestion
    }
    throw new Error("Invariant: unknown line type")
```

Reply:
Added in [d76a9f3](https://github.com/saleor/apps/pull/1644/commits/d76a9f31bcb83eedcbf352d5bc5022171d91bc32)